packages/atomic/src/utils/initialization-utils.tsx

```diff
@@ -1,53 +1,132 @@
-import {Engine} from '@coveo/headless';
+import {Controller, Engine} from '@coveo/headless';
 import {ComponentInterface, getElement, h} from '@stencil/core';
 import {i18n} from 'i18next';
 
-export interface InterfaceContext {
+/**
+ * Bindings passed from the `AtomicSearchInterface` to its children components.
+ */
+export interface Bindings {
+  /**
+   * Headless Engine instance.
+   */
   engine: Engine;
+  /**
+   * i18n instance, for localization.
+   */
   i18n: i18n;
 }
 
-export type InitializeEventHandler = (context: InterfaceContext) => void;
+export type InitializeEventHandler = (bindings: Bindings) => void;
 export type InitializeEvent = CustomEvent<InitializeEventHandler>;
 
-export class InitializationError extends Error {
+export class MissingInterfaceParentError extends Error {
   constructor(elementName: string) {
     super(
       `The "${elementName}" element must be the child of a configured "atomic-search-interface" element.`
     );
-    this.name = 'InitializationError';
   }
 }
 
-export function Initialization() {
-  return (component: ComponentInterface, initializeMethod: string) => {
+/**
+ * Necessary interface an Atomic Component must have to initialize itself correctly.
+ */
+export interface AtomicComponentInterface extends ComponentInterface {
+  /**
+   * Bindings passed from the `AtomicSearchInterface` to its children components.
+   */
+  bindings: Bindings;
+  /**
+   * Record of methods earch returning an i18n localized string.
+   */
+  strings?: Record<string, () => string>;
+  /**
+   * Headless Controller instance associated with the Atomic Component.
+   */
+  controller?: Controller;
+  /**
+   * Headless Controller's state.
+   */
+  controllerState?: unknown;
+  /**
+   * Callback for when the subscribe method is called and the controller's state is updated.
+   */
+  onControllerStateUpdate?: () => void;
+  /**
+   * Error that, when defined, will be rendered inside an `atomic-component-error` component.
+   */
+  error?: Error;
+}
+
+/**
+ * Utility that automatically fetches the `bindings` from the parent `atomic-search-interface` component. This decorator should be applied to the `initialize` method directly.
+ *
+ * In order for a component using this decorator to render properly, it should have an internal state property using data from the `bindings`. For more information, view the "Utilities" section of the readme.
+ */
+export function Initialization(options?: {
+  resubscribeControllerOnConnectedCallback?: boolean;
+}) {
+  return (component: AtomicComponentInterface, initializeMethod: string) => {
     const {
+      connectedCallback,
       componentWillLoad,
       render,
       componentDidRender,
       componentDidLoad,
-      [initializeMethod]: initialize,
+      disconnectedCallback,
+      onControllerStateUpdate,
     } = component;
-    const errorProperty = 'error';
-    const contextProperty = 'context';
+    const initialize: () => void = component[initializeMethod];
```

Comment:
Stylistic change, but just breaking up this function into: 

```
component.connectedCallback = buildConnectedCallback(this);
component.componentWillLoad = buildComponentWillLoad(this);
component.render= buildRender(this);

// etc.. for each component property
// each build function return a function bound to "this" passed as parameter
```

I think this would make it more easily readable as separate logical unit.

Reply:
Would also probably make each of these function more easily unit testable individually ?